---
title: "TCGA analysis for Braun et al., manuscript"
author: "Dillon Corvino"
date: "11/06/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Environment Set up
#rm(list = ls()) #Clean workspace
#cat("\014")     #Clean Console

require(rstudioapi)
require(ggplot2)
require(gplots)
require(TCGAbiolinks)
require(SummarizedExperiment)
require(car)
require(psych)


#Set wd to source file location
#setwd(dirname(getActiveDocumentContext()$path))
#getwd()


#source("https://bioconductor.org/biocLite.R")
#biocLite("TCGAbiolinks")

# If GDC server down, may need to update TCGAbiolinks, run the below code and RESTART R
#devtools::install_github(repo = "BioinformaticsFMRP/TCGAbiolinks",dependencies = T)

#source('https://bioconductor.org/biocLite.R')
#BiocManager::install("remotes")
#BiocManager::install('Bioconductor/GenomicDataCommons')
#BiocManager::install('BioinformaticsFMRP/TCGAbiolinks',ref = "GenomicDataCommons")

library("GenomicDataCommons")

GenomicDataCommons::status()

if(!dir.exists("Output")){dir.create("Output")}
if(!dir.exists("Output/Data")){dir.create("Output/Data")}

```

```{r Moving_average_analysis}

#############################################
# Moving average analayis was adapted from:
# Riesenberg, S. et al. MITF and c-Jun antagonism interconnects melanoma dedifferentiation with pro-inflammatory cytokine responsiveness and myeloid cell recruitment. Nat. Commun. 6:8755 doi: 10.1038/ncomms9755 (2015).
#############################################

# Required packages for Moving average analysis 
library("Biobase")
library("cgdsr")

# Function for moving average analysis
#  see also http://www.cookbook-r.com/
#  free usage under creative commons licence http://creativecommons.org/licenses/by-sa/3.0/legalcode

# Moving average function
movingAverage <- function(x, n=1, centered=FALSE) {
  
  if (centered) {
    before <- floor  ((n-1)/2)
    after  <- ceiling((n-1)/2)
  } else {
    before <- n-1
    after  <- 0
  }
  
  # Track the sum and count of number of non-NA items
  s     <- rep(0, length(x))
  count <- rep(0, length(x))
  
  # Add the centered data 
  new <- x
  # Add to count list wherever there isn't a 
  count <- count + !is.na(new)
  # Now replace NA_s with 0_s and add to total
  new[is.na(new)] <- 0
  s <- s + new
  
  # Add the data from before
  i <- 1
  while (i <= before) {
    # This is the vector with offset values to add
    new   <- c(rep(NA, i), x[1:(length(x)-i)])
    
    count <- count + !is.na(new)
    new[is.na(new)] <- 0
    s <- s + new
    
    i <- i+1
  }
  
  # Add the data from after
  i <- 1
  while (i <= after) {
    # This is the vector with offset values to add
    new   <- c(x[(i+1):length(x)], rep(NA, i))
    
    count <- count + !is.na(new)
    new[is.na(new)] <- 0
    s <- s + new
    
    i <- i+1
  }
  
  # return sum divided by count
  s/count
}

# Which TCGA datasets to query 
movingaverage.datasets <- c("skcm_tcga", "hnsc_tcga")

# Calculate and plot moving average for selected datasets
for(dataset in 1:length(movingaverage.datasets)){
  
  # Initialise which TCGA dataset to use
  tcga.cancer <- paste0(movingaverage.datasets[dataset])
  
  
  # Create output directory
  output.dir <- paste0("Output/Moving_average_", tcga.cancer)
  
  if(!dir.exists(paste0(output.dir))){
    dir.create(paste0(output.dir), recursive = TRUE)}
  
  
  # Get tcga data using CGDS R package
  # create CGDS object and connection to cBioportal
  mycgds = CGDS("http://www.cbioportal.org/", verbose = T)
  
  # test Bioportal connection
  test(mycgds)
  
  # check list of cancer studies at server 
  getCancerStudies(mycgds)[,1:2]
  
  # check status and available data
  getGeneticProfiles(mycgds, tcga.cancer)[,1:2]
  getCaseLists(mycgds, tcga.cancer)[,1:2]
  
  # Get gene expression data using getProfileData
  genetic_profile_id <-paste0(tcga.cancer,"_rna_seq_v2_mrna")
  case_list_id <- paste0(tcga.cancer, "_rna_seq_v2_mrna")
  
  
  # enter genes of interest to retrieve via cBioportal
  # Vector of genes of interest
  gene <- c("CD226", "CD8B", "NCAM1", "IFNG", "GZMB", "PVR", "NECTIN2")
  
  # get gene expression data frame 
  gedf <- getProfileData(mycgds, gene, genetic_profile_id, case_list_id)
  
  # check normalized gene expression data (read counts) and log2 transform (avoid negative values)
  dim(gedf)
  ncol(gedf) == length(gene) # TRUE
  head(gedf)
  gedf[gedf < 1] <- 1
  gedf <- log2(gedf)
  head(gedf)
  n.val <- nrow(gedf)
  
  # define gene used for ranking of samples
  xSortGene <- gedf$CD226
  xSortGeneID <- "CD226"
  
  xsort <- sort(xSortGene, decreasing = FALSE, index.return = TRUE)
  
  # Plot moving average for all GOI except first (CD226)
  for(gene.i in 2:length(gene)){
    
    # define dependent gene of interest
    xDepGene <- gedf[ , colnames(gedf) == gene[gene.i]]
    xDepGeneID <- gene[gene.i]
    
    xLeftGene <- xDepGene[xsort$ix] - median(xDepGene[xsort$ix])
    
    
    par(mar = c(5.1, 4.1, 4.1, 4.2))
    
    # Set plot params
    sample.width = 20 # change sample window size for moving average calculation
    laxis.col = "blue" # change colour of dependent gene
    x <- xLeftGene
    
    ylim.plot <- c(summary(xLeftGene)[1], summary(xLeftGene)[6]) # change scale of dependent gene y-axis
    
    # Not for visualisation purposes - the yaxis range for NECTIN2 and PVR were altered for SKCM dataset.
    
    
    # Plot bars of dependent gene expression as a background trace
    plot(x, col = "lightgrey", type = "h", lwd = 0.5, ylim = ylim.plot, xlab = "", ylab ="", axes = F)
    box(which = "plot")
    axis(side = 1, outer = F)
    
    # Plot moving average line for the dependent gene
    axis(side = 2, col = laxis.col, col.axis = laxis.col, lwd = 2)
    abline(h = 0, col = "lightgrey", lwd = 0.5)
    lines(movingAverage(x, n = sample.width, centered =T), col = laxis.col, type = "l", lwd = 3)
    par(new=T)
    
    # Plot line for the ranking gene 
    plot(xsort$x - median(xsort$x), col = "black", type = "l", lwd = 3, xlab = "", ylab ="", axes = F)
    axis(side = 4, col = "black", col.axis = "black", lwd = 2)
    
    # Add axes labels
    title(main = paste0(xDepGeneID, " in ", tcga.cancer),
          col.main = laxis.col,
          xlab = "TCGA samples ranked by CD226 expression",
          ylab = paste0("Moving average ", xDepGeneID),
          col.lab = laxis.col)
    
    mtext(paste0(xSortGeneID, " expression (log2)"), side = 4, line = 2.3)
    
    
    # Correlation values
    cor.val <- cor(xLeftGene, xsort$x, method = "spearman")
    cor.val <- signif(cor.val, 4)
    mtext(paste0("r = ", cor.val), adj = 0)
    
    
    cor.test.val <- cor.test(xLeftGene, xsort$x, method = "spearman")
    corr.pval <- signif(cor.test.val$p.value, 4)
    mtext(paste0("P = ", corr.pval), adj = 1)
    
    mtext(paste0("n = ", n.val), adj = 0.5)
    
    # Save plot to file
    dev.copy(pdf, paste0(output.dir, "/Moving_average_CD226_vs_", xDepGeneID, ".pdf"))
    dev.off()
    
  }
  
  
  
}

```


